Created test script for models

kelaaron opened the pull request:
Created test script for testing certain metrics for models, located at `../inference/models/tests`.

To call the test script, navigate to that directory and call `python3 models_test.py MODEL TESTCRITERIA PATH_TO_TEST_CASE`. Currently, the only model supported is "classifier" and the only testcriteria is "accuracy", however other models and classifiers should be easy to add in the future.

Let me know if anything should be changed / added! 😁

skekre98 commented:
Awesome, this is way better than what I had in mind! I had a few design tweaks that I think might work well if you're willing to add them into this PR. If not, I can go ahead and merge it as is 😄 .

kelaaron commented:
Sure - I can probably add them to this PR, what changes were you thinking?

skekre98 commented:
> Sure - I can probably add them to this PR, what changes were you thinking?

I like the idea of using model type in the test script. However in the future we may have multiple classifiers being used. So instead of hardcoding the name of the model, perhaps we could create a separate directory called `classifiers` and test all of them when the classifier flag is used. We'd also need to map the test files to each model so maybe use a python dictionary for that

Ex.
```python
model_test_map = {
    "model1" : "../path/to/test/file1",
    "model2" : "../path/to/test/file2",
    "model3" : "../path/to/test/file3"
}

What do you think?

kelaaron commented:
> > Sure - I can probably add them to this PR, what changes were you thinking?
> 
> I like the idea of using model type in the test script. However in the future we may have multiple classifiers being used. So instead of hardcoding the name of the model, perhaps we could create a separate directory under `models` called `classifiers` and test all of them when the classifier flag is used. We'd also need to map the test files to each model so maybe use a python dictionary for that
> 
> Ex.
> 
> ```python
> model_test_map = {
>     "model1" : "../path/to/test/file1",
>     "model2" : "../path/to/test/file2",
>     "model3" : "../path/to/test/file3"
> }
> ```
> 
> What do you think?

That's a great idea! Just to clarify - are you saying I should move 'query_classifier.py' into models/classifiers? I'm thinking doing so might potentially cause other files to break.

skekre98 commented:
@kelaaron I agree, I can handle those changes on my end before we merge this PR. Let's go ahead and move `query_classifier.pkl` into `../inference/models/classifers/`

kelaaron commented:
@skekre98 I added the model map and moved `query_classifier.pkl` into `../inference/models/classifers/`, I think it's definitely easier to modify now! Let me know if there's any other changes I should make. 😁

skekre98 commented on inference/models/tests/models_test.py:
```diff
@@ -0,0 +1,56 @@
+    for model, test_path in model_test_map.items():
+
+
+        os.system('python3 ' + model) # train models, dumps into model .pkl file
```
There is nothing wrong with this implementation, but I would suggest importing query classifier as a module so you won't be spawning a new process for every model. There are other instances where `os.system`  is being used in the repository, and I have been meaning to update those to be modules as well 😅 . This script will likely be used when performing pre-deployment checks so it would be a good idea to be efficient with compute.

It would look something like this:
```python
import query_classifier

query_classifier.main()
```

skekre98 reviewed (approved):
@kelaaron really great work 🎉 , this has been an issue for quite sometime. Thanks for getting it done! I have added a minor comment if you were interested in looking into it in this PR, otherwise I can go ahead and merge it and handle it in a future PR. I was going to open up an issue as is to remove the use of `os.system` and convert to using modules. Let me know how you would like me to handle this.

kelaaron commented:
> @kelaaron really great work 🎉 , this has been an issue for quite sometime. Thanks for getting it done! I have added a minor comment if you were interested in looking into it in this PR, otherwise I can go ahead and merge it and handle it in a future PR. I was going to open up an issue to remove the use of `os.system` and convert to using modules anyways. Let me know how you would like me to handle this.

@skekre98 I just pushed a fix which utilizes getattr to utilize each model's main method if we import them. Let me know if this works for you!

skekre98 reviewed (approved):
@kelaaron really great work 😄
